ext/random/random.c

```diff
@@ -680,35 +680,82 @@ uint64_t php_random_generate_fallback_seed(void)
 		gettimeofday(&tv, NULL);
 		fallback_seed_add(&c, &tv, sizeof(tv));
 		/* Previous state. */
-		fallback_seed_add(&c, RANDOM_G(fallback_seed), 20);
+		fallback_seed_add(&c, state->seed, 20);
 	}
-	PHP_SHA1Final(RANDOM_G(fallback_seed), &c);
-	RANDOM_G(fallback_seed_initialized) = true;
+	PHP_SHA1Final(state->seed, &c);
+	state->initialized = true;
 
 	uint64_t result = 0;
 
 	for (int i = 0; i < sizeof(result); i++) {
-		result = result | (((uint64_t)RANDOM_G(fallback_seed)[i]) << (i * 8));
+		result = result | (((uint64_t)state->seed[i]) << (i * 8));
 	}
 
 	return result;
 }
 
-/* {{{ PHP_GINIT_FUNCTION */
-static PHP_GINIT_FUNCTION(random)
+uint64_t php_random_generate_fallback_seed(void)
 {
-	random_globals->random_fd = -1;
-	random_globals->fallback_seed_initialized = false;
+	return php_random_generate_fallback_seed_ex(&RANDOM_G(fallback_seed_state));
 }
-/* }}} */
 
-/* {{{ PHP_GSHUTDOWN_FUNCTION */
-static PHP_GSHUTDOWN_FUNCTION(random)
+static zend_result php_general_random_bytes_for_zend_initialize(php_random_state_for_zend *state)
+{
+	uint64_t t[4];
+
+	do {
+		char errstr[128];
+		if (php_random_bytes_ex(&t, sizeof(t), errstr, sizeof(errstr)) == FAILURE) {
+#if ZEND_DEBUG
+			fprintf(stderr, "php_random_bytes_ex: Failed to generate a random seed: %s\n", errstr);
+#endif
+			return FAILURE;
+		}
+	} while (UNEXPECTED(t[0] == 0 && t[1] == 0 && t[2] == 0 && t[3] == 0));
+
+	php_random_xoshiro256starstar_seed256(&state->xoshiro256starstar_state, t[0], t[1], t[2], t[3]);
+
+	return SUCCESS;
+}
+
+static void php_general_random_bytes_for_zend_initialize_fallback(php_random_state_for_zend *state)
+{
+	uint64_t t;
+	php_random_fallback_seed_state fallback_state;
+
+	do {
+		t = php_random_generate_fallback_seed_ex(&fallback_state);
+	} while (UNEXPECTED(t == 0));
+
+	php_random_xoshiro256starstar_seed64(&state->xoshiro256starstar_state, t);
+}
```

Comment:
The `fallback_state` is not initialized, thus reading unspecified stack memory.

Furthermore using `seed64` already guarantees that the state will not be all-zeroes, making the do-while loop unnecessary.

However even when fixing this, the implementation is broken: It uses 64 bits of randomness to fill the entire 256-bit state of xoshiro256**. Depending on what information is available, the fallback generator is capable of using providing more than 64 bits of entropy. Thus the intended use is to call the fallback generator as often as necessary.

I would suggest the following (untested). It minimizes the amount of code that needs to be repeated and takes care to not call the CSPRNG once it determined that the CSPRNG is broken:

```c
static void php_general_random_bytes_for_zend_initialize(php_random_state_for_zend *state)
{
	uint64_t t[4];
	php_random_fallback_seed_state fallback_state = {
		.initialized = false
	};

	do {
		if (!fallback_state.initialized) {
			char errstr[128];
			if (php_random_bytes_ex(&t, sizeof(t), errstr, sizeof(errstr)) == FAILURE) {
#if ZEND_DEBUG
				fprintf(stderr, "php_random_bytes_ex: Failed to generate a random seed: %s\n", errstr);
#endif
				goto fallback;
			}
		} else {
 fallback:
			t[0] = php_random_generate_fallback_seed_ex(&fallback_state);
			t[1] = php_random_generate_fallback_seed_ex(&fallback_state);
			t[2] = php_random_generate_fallback_seed_ex(&fallback_state);
			t[3] = php_random_generate_fallback_seed_ex(&fallback_state);
		}
	} while (UNEXPECTED(t[0] == 0 && t[1] == 0 && t[2] == 0 && t[3] == 0));

	php_random_xoshiro256starstar_seed256(&state->xoshiro256starstar_state, t[0], t[1], t[2], t[3]);
}
```